rust-symcrypt/src/cipher/cbc.rs

```diff
@@ -0,0 +1,319 @@
+//! Functions for Cbc encryption and decryption. For more info please see symcrypt.h
+//!
+//! These functions are methods on the [AesExpandedKey] struct.
+//!
+//! # Examples
+//!
+//! ## AES-CBC Encryption in one go
+//!
+//! ```rust
+//! use symcrypt::cipher::AesExpandedKey;
+//! use hex::*;
+//!
+//! // Set up the key, chaining value, and plaintext
+//! let key = hex::decode("00000000000000000000000000000000").unwrap();
+//! let aes_cbc = AesExpandedKey::new(&key).unwrap();
+//! let mut chaining_value = hex::decode("00000000000000000000000000000000").unwrap();
+//! let mut plain_text = hex::decode("f34481ec3cc627bacd5dc3fb08f273e6").unwrap();
+//! let mut cipher_text = vec![0u8; plain_text.len()];
+//!
+//! // Encrypt the plaintext
+//! aes_cbc.aes_cbc_encrypt(&mut chaining_value, &plain_text, &mut cipher_text).unwrap();
+//!
+//! assert_eq!(hex::encode(cipher_text), "0336763e966d92595a567cc9ce537f5e");
+//! ```
+//!
+//! ## AES-CBC Encryption piecewise
+//!
+//! ```rust
+//! use symcrypt::cipher::AesExpandedKey;
+//! use hex::*;
+//!
+//! // Set up the key, chaining value, and plaintext
+//! let key = hex::decode("5d98398b5e3b98d87e07ecf1332df4ac").unwrap();
+//! let mut chaining_value = hex::decode("db22065fb9302c4445151adc91310797").unwrap();
+//! let plaintext = hex::decode("4831f8d1a92cf167a444ccae8d90158dfc55c9a0742019e642116bbaa87aa205").unwrap();
+//! let expected_ciphertext = hex::decode("f03f86e1a6f1e23e70af3f3ab3b777fd43103f2e7a6fc245a3656799176a2611").unwrap();
+//!
+//! // Create the AES key
+//! let aes_cbc = AesExpandedKey::new(&key).unwrap();
+//!
+//! // Setup piecewise encrypt
+//! let mut piecewise_encrypted = vec![0u8; plaintext.len()];
+//! let mid = plaintext.len() / 2;
```

Comment:
Not a real pattern.

Better pattern is an array of fixed disk blocks to encrypt/decrypt.